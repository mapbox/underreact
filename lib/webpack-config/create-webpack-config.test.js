'use strict';

const path = require('path');
const tempy = require('tempy');
const fs = require('fs');
const browserslist = require('browserslist');

jest.mock('../config/get-user-config');

const getUserConfig = require('../config/get-user-config');
const config = require('../config');
const createWebpackConfig = require('./create-webpack-config');

const getCliOpts = ({
  command = 'build',
  mode = 'development',
  configPath = path.join(process.cwd(), 'underreact.config.js')
} = {}) => ({
  command,
  mode,
  configPath
});

test('Basic Test', () => {
  getUserConfig.mockResolvedValueOnce({
    siteBasePath: 'fancy',
    publicAssetsPath: 'cacheable-things',
    htmlSource: `<html></html>`,
    webpackLoaders: [
      {
        test: /\.less$/,
        use: ['css-loader', 'less-loader']
      }
    ],
    webpackPlugins: [function foo() {}],
    webpackConfigTransform: config => {
      config.devtool = false;
      return config;
    }
  });

  const urcPromise = config(getCliOpts());

  return expect(
    urcPromise.then(urc => createWebpackConfig(urc))
  ).resolves.toMatchSnapshot();
});

test('Adds mapbox-gl to exclude if user is on 2.0.0 or above', () => {
  getUserConfig.mockResolvedValueOnce({});
  const tempDir = tempy.directory();
  fs.writeFileSync(
    path.join(tempDir, 'package.json'),
    `{
      "name": "fake",
      "version": "0.0.0",
      "dependencies": {
        "mapbox-gl": "^2.0.0"
      }
    }`
  );

  fs.writeFileSync(
    path.join(tempDir, 'babel.config.js'),
    `module.exports = {}`
  );

  const urcPromise = config(
    getCliOpts({
      configPath: path.join(tempDir, 'underreact.config.js')
    })
  );

  return urcPromise.then(urc => {
    const config = createWebpackConfig(urc);
    expect(config.module.rules[0].oneOf[1].include).toEqual(
      /\/node_modules\/((?!mapbox-gl\/))/
    );
  });
});

test('Does not add mapbox-gl to exclude if user is on a version below 2.0.0', () => {
  getUserConfig.mockResolvedValueOnce({});
  const tempDir = tempy.directory();
  fs.writeFileSync(
    path.join(tempDir, 'package.json'),
    `{
      "name": "fake",
      "version": "0.0.0",
      "dependencies": {
        "mapbox-gl": "~1.9.0"
      }
    }`
  );

  fs.writeFileSync(
    path.join(tempDir, 'babel.config.js'),
    `module.exports = {}`
  );

  const urcPromise = config(
    getCliOpts({
      configPath: path.join(tempDir, 'underreact.config.js')
    })
  );

  return urcPromise.then(urc => {
    const config = createWebpackConfig(urc);
    expect(config.module.rules[0].oneOf[1].include).toBeUndefined();
  });
});

test('Uses babel.config.js if it exists at project root', () => {
  getUserConfig.mockResolvedValueOnce({});
  const tempDir = tempy.directory();
  const urcPromise = config(
    getCliOpts({
      configPath: path.join(tempDir, 'underreact.config.js')
    })
  );

  fs.writeFileSync(
    path.join(tempDir, 'babel.config.js'),
    `module.exports = {}`
  );

  return urcPromise.then(urc => {
    expect(
      createWebpackConfig(urc).module.rules[0].oneOf.find(
        obj => obj.test.toString() === `/\\.jsx?$/`
      ).options.configFile
    ).toBe(path.join(tempDir, 'babel.config.js'));
  });
});

test('Uses babel preset if babel.config.js doesnt exists at project root', () => {
  getUserConfig.mockResolvedValueOnce({});
  const tempDir = tempy.directory();
  const urcPromise = config(
    getCliOpts({
      configPath: path.join(tempDir, 'underreact.config.js')
    })
  );
  return urcPromise.then(urc => {
    expect(
      createWebpackConfig(urc).module.rules[0].oneOf.find(
        obj => obj.test.toString() === `/\\.jsx?$/`
      ).options.presets[0]
    ).toBe(require('@mapbox/babel-preset-mapbox'));
  });
});

test('babel-loader creates a good cacheIdentifier which includes urc.browserslist', () => {
  const query = 'chrome 67';
  getUserConfig.mockResolvedValueOnce({ browserslist: [query] });

  const urcPromise = config(
    getCliOpts({
      configPath: '/fake/underreact.config.js'
    })
  );

  return urcPromise.then(urc => {
    const cacheIdentifier = JSON.parse(
      createWebpackConfig(urc).module.rules[0].oneOf.find(
        obj => obj.test.toString() === `/\\.jsx?$/`
      ).options.cacheIdentifier
    );

    expect(cacheIdentifier).toMatchObject({
      'babel-loader': expect.any(String),
      '@babel/core': expect.any(String),
      babelConfig: expect.any(String),
      env: 'test',
      browserslist: browserslist(query)
    });
  });
});

test('config transform works', () => {
  getUserConfig.mockResolvedValueOnce({
    webpackConfigTransform: obj =>
      Object.assign({ injectedObj: 'injectedObj' }, obj)
  });

  const urcPromise = config(
    getCliOpts({
      configPath: path.join(process.cwd(), 'underreact.config.js')
    })
  );
  return urcPromise.then(urc => {
    expect(createWebpackConfig(urc)).toMatchObject({
      injectedObj: 'injectedObj'
    });
  });
});

test("Doesn't set polyfill entry when urc.polyfill is false", () => {
  getUserConfig.mockResolvedValueOnce({
    polyfill: false
  });

  const urcPromise = config(
    getCliOpts({
      configPath: path.join(process.cwd(), 'underreact.config.js')
    })
  );
  return urcPromise.then(urc => {
    expect(createWebpackConfig(urc).entry.polyfill).toBe(undefined);
  });
});
