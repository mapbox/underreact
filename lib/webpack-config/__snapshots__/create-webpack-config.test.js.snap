// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Test 1`] = `
Object {
  "bail": false,
  "context": "<PROJECT_ROOT>",
  "devtool": false,
  "entry": Object {
    "main": Array [
      "<PROJECT_ROOT>/src/index.js",
    ],
    "polyfill": "<PROJECT_ROOT>/polyfill/index.js",
    "vendor": Array [
      "<PROJECT_ROOT>/node_modules/react",
      "<PROJECT_ROOT>/node_modules/react-dom",
    ],
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "test": /\\\\\\.less\\$/,
            "use": Array [
              "css-loader",
              "less-loader",
            ],
          },
          Object {
            "exclude": Array [
              /\\[/\\\\\\\\\\\\\\\\\\]node_modules\\[/\\\\\\\\\\\\\\\\\\]/,
            ],
            "loader": "<PROJECT_ROOT>/node_modules/babel-loader/lib/index.js",
            "options": Object {
              "babelrc": false,
              "cacheCompression": false,
              "cacheDirectory": true,
              "cacheIdentifier": "{\\"babel-loader\\":\\"8.2.2\\",\\"@babel/core\\":\\"7.12.10\\",\\"babelConfig\\":\\"\\",\\"env\\":\\"test\\",\\"browserslist\\":[\\"and_chr 87\\",\\"and_ff 83\\",\\"and_uc 12.12\\",\\"chrome 87\\",\\"chrome 86\\",\\"chrome 85\\",\\"chrome 84\\",\\"chrome 49\\",\\"edge 87\\",\\"firefox 84\\",\\"firefox 83\\",\\"ie 11\\",\\"ios_saf 14.0-14.3\\",\\"ios_saf 13.4-13.7\\",\\"ios_saf 13.3\\",\\"ios_saf 12.2-12.4\\",\\"opera 72\\",\\"safari 14\\",\\"safari 13.1\\",\\"samsung 13.0\\",\\"samsung 12.0\\"]}",
              "compact": false,
              "configFile": false,
              "presets": Array [
                [Function],
              ],
            },
            "test": /\\\\\\.jsx\\?\\$/,
          },
          Object {
            "exclude": /@babel\\(\\?:\\\\/\\|\\\\\\\\\\{1,2\\}\\)runtime/,
            "loader": "<PROJECT_ROOT>/node_modules/babel-loader/lib/index.js",
            "options": Object {
              "babelrc": false,
              "cacheCompression": false,
              "cacheDirectory": true,
              "cacheIdentifier": "{\\"babel-loader\\":\\"8.2.2\\",\\"@babel/core\\":\\"7.12.10\\",\\"babelConfig\\":\\"\\",\\"env\\":\\"test\\",\\"browserslist\\":[\\"and_chr 87\\",\\"and_ff 83\\",\\"and_uc 12.12\\",\\"chrome 87\\",\\"chrome 86\\",\\"chrome 85\\",\\"chrome 84\\",\\"chrome 49\\",\\"edge 87\\",\\"firefox 84\\",\\"firefox 83\\",\\"ie 11\\",\\"ios_saf 14.0-14.3\\",\\"ios_saf 13.4-13.7\\",\\"ios_saf 13.3\\",\\"ios_saf 12.2-12.4\\",\\"opera 72\\",\\"safari 14\\",\\"safari 13.1\\",\\"samsung 13.0\\",\\"samsung 12.0\\"]}",
              "compact": false,
              "configFile": false,
              "presets": Array [
                Array [
                  [Function],
                  Object {
                    "override": [Function],
                  },
                ],
              ],
              "sourceMaps": false,
              "sourceType": "unambiguous",
            },
            "test": /\\\\\\.js\\$/,
          },
          Object {
            "test": /\\\\\\.css\\$/,
            "use": Array [
              "<PROJECT_ROOT>/node_modules/style-loader/index.js",
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/css-loader/index.js",
                "options": Object {
                  "importLoaders": 1,
                  "sourceMap": true,
                },
              },
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/postcss-loader/lib/index.js",
                "options": Object {
                  "ident": "postcss",
                  "plugins": Array [
                    [Function],
                  ],
                },
              },
            ],
          },
          Object {
            "issuer": /\\\\\\.\\(js\\|jsx\\)\\$/,
            "test": /\\\\\\.svg\\$/,
            "use": Object {
              "loader": "@mapbox/svg-react-transformer-loader",
              "options": Object {
                "template": "fancy",
              },
            },
          },
          Object {
            "exclude": Array [
              /\\\\\\.\\(js\\|jsx\\)\\$/,
              /\\\\\\.template\\$/,
              /\\\\\\.json\\$/,
              /\\\\\\.ejs\\$/,
            ],
            "loader": "<PROJECT_ROOT>/node_modules/file-loader/dist/cjs.js",
            "options": Object {
              "name": "cacheable-things/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "child_process": "empty",
    "dgram": "empty",
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "minimizer": Array [
      TerserPlugin {
        "options": Object {
          "cache": true,
          "cacheKeys": [Function],
          "chunkFilter": [Function],
          "exclude": undefined,
          "extractComments": false,
          "include": undefined,
          "minify": undefined,
          "parallel": true,
          "sourceMap": true,
          "terserOptions": Object {
            "compress": Object {
              "comparisons": false,
              "ecma": 5,
              "inline": 2,
              "warnings": false,
            },
            "mangle": Object {
              "safari10": true,
            },
            "output": Object {
              "ascii_only": true,
              "comments": false,
              "ecma": 5,
            },
            "parse": Object {
              "ecma": 8,
            },
          },
          "test": /\\\\\\.m\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
          "warningsFilter": [Function],
        },
      },
    ],
    "runtimeChunk": "single",
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendor": Object {
          "chunks": "initial",
          "enforce": true,
          "name": "vendor",
          "test": "vendor",
        },
      },
    },
  },
  "output": Object {
    "chunkFilename": "cacheable-things/js/[name].chunk.js",
    "filename": "cacheable-things/js/[name].js",
    "path": "<PROJECT_ROOT>/_site",
    "publicPath": "/fancy/",
  },
  "performance": false,
  "plugins": Array [
    AssetsWebpackPlugin {
      "options": Object {
        "fileTypes": Array [
          "js",
          "css",
        ],
        "filename": "assets.json",
        "fullPath": true,
        "includeAllFileTypes": true,
        "integrity": false,
        "keepInMemory": false,
        "manifestFirst": true,
        "path": "<PROJECT_ROOT>/_site",
        "prettyPrint": false,
        "processOutput": [Function],
        "update": false,
        "useCompilerPath": false,
      },
      "writer": [Function],
    },
    EnvironmentPlugin {
      "defaultValues": Object {
        "BASE_PATH": "/fancy",
        "DEPLOY_ENV": "development",
        "NODE_ENV": "test",
      },
      "keys": Array [
        "NODE_ENV",
        "DEPLOY_ENV",
        "BASE_PATH",
      ],
    },
    MiniCssExtractPlugin {
      "options": Object {
        "chunkFilename": "cacheable-things/css/[name]-[contenthash:8].chunk.css",
        "filename": "cacheable-things/css/[name]-[contenthash:8].css",
      },
    },
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunks": "all",
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [
          "polyfill",
          "runtime",
        ],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": true,
        "meta": Object {},
        "minify": Object {},
        "showErrors": true,
        "template": "<PROJECT_ROOT>/node_modules/html-webpack-plugin/default_index.ejs",
        "templateContent": [Function],
        "templateParameters": [Function],
        "title": "Webpack App",
        "xhtml": false,
      },
    },
    [Function],
    ExcludeChunksPlugin {
      "excludeChunks": Array [
        "runtime",
      ],
    },
  ],
  "resolve": Object {
    "alias": Object {
      "@babel/runtime": "<PROJECT_ROOT>/node_modules/@babel/runtime",
    },
  },
  "resolveLoader": Object {
    "modules": Array [
      "<PROJECT_ROOT>/node_modules",
      "node_modules",
    ],
  },
}
`;
